### DISCUSSION TITLE
Trying to add a feature but GtkScale won't work correctly.

### DISCUSSION THREAD

**Discussion by GithubAnon0000** [xournalpp/xournalpp#5842] [Q&A] [answered]
Hello!

I am attempting to add #1962 and started in my [own repo](https://github.com/GithubAnon0000/xournalpp_add_default_page_zoom/pull/2). The slider shows in the settings menu and the value that is selected will be saved in the settings xml file. The slider won't move though, which means the user has no indication that this actually worked.

I only have little experience with coding in general (with gtk and C). But in C this task is fairly easy:

<details><summary>C code example</summary>
<p>

```c
// Compile with: gcc main.c -O3 -Wall -Wextra -Wno-unused $(pkg-config --libs --cflags gtk+-3.0) -o main && ./main
#include <gtk/gtk.h>




static void
activate(GtkApplication *application, gpointer user_data);




int
main(void)
{
	GtkApplication *application;
	int status;
	
	application = gtk_application_new("org.Test",
	                                  G_APPLICATION_FLAGS_NONE);
	
	g_signal_connect(application, "activate", G_CALLBACK(activate), NULL);
	status = g_application_run(G_APPLICATION(application), 0, NULL);
	
	g_object_unref(application);
	
	return status;
} /* main */




static void
activate(GtkApplication *application, gpointer user_data)
{
	g_return_if_fail(GTK_IS_APPLICATION(application));
	
	
	// app window
	GtkWidget *window;
	int window_width;
	int window_height;
	
	window_width  = 400;
	window_height =  80;
	
	window = gtk_application_window_new(application);
	
	gtk_window_set_default_size(GTK_WINDOW(window), window_width,
	                            window_height);
	
	
	
	
	// container
	GtkWidget *box;
	
	box = gtk_box_new(GTK_ORIENTATION_HORIZONTAL, 5);
	gtk_container_add(GTK_CONTAINER(window), box);
	
	
	
	
	// slider
	GtkAdjustment *adjustment;
	GtkWidget     *scale;
	
	adjustment = gtk_adjustment_new(100.00, 30.00, 700.00, 1.00, 0.00, 0.00);
	
	scale = gtk_scale_new(GTK_ORIENTATION_HORIZONTAL,
	                      GTK_ADJUSTMENT(adjustment));
	
	gtk_box_pack_start(GTK_BOX(box), scale, TRUE, TRUE, 0);
	
	
	
	// show all widgets
	gtk_widget_show_all(window);
} /* activate */
``` 

</p>
</details> 

This works and the slider can be moved easily. I noticed that if I add `zoomcallib_set_val(ZOOM_CALLIB(callib), dpi);` to `src/core/gui/dialog/SettingsDialog.cpp` into the `setDefaultPageZoom` function, the slider _can_ be moved (but then affects the dpi scaling which we obviously do not want).  Looking at `ZoomCallib.cpp` / `ZoomCallib.h` I don't see what makes it work though.

So in short: I need help getting this slider to move properly. After that I'll see how to apply the default zoom effect.

Any help is appreciated!

**Comment by bhennion** [comment, marked as answer]
The issue is your signal handler:
```
    g_signal_connect(builder.get("sidDefaultZoomSlider"), "change-value",
                    G_CALLBACK(+[](GtkRange*, GtkScrollType, gdouble value, SettingsDialog* self) {
                        self->setDefaultPageZoom(round_cast<int>(value));
                    }),
                    this);
```
If you read the [documentation](https://docs.gtk.org/gtk3/signal.Range.change-value.html) on the signal "change-value", you'll see that the callback should return a boolean: true if no further processing is required, false if other callbacks should be called afterwards.
Now, the cursor is actually moved by GTK's default callback, so your callback must return `false`, otherwise the cursor isn't moved.

```
    g_signal_connect(builder.get("sidDefaultZoomSlider"), "change-value",
                    G_CALLBACK(+[](GtkRange*, GtkScrollType, gdouble value, SettingsDialog* self) {
                        self->setDefaultPageZoom(round_cast<int>(value));
                        return false;
                    }),
                    this);
```
Without the `return false;` statement, you have an undefined behaviour (UB): a pointer to a function of type void(Args...) is cast to a pointer to a function gboolean(Args...) so you have no guaranti on the returned value. Using another compiler or another version of the same compiler could alter the behaviour.

Now, strictly speaking, you should even do the pedantic, to avoid any UB
```
    g_signal_connect(builder.get("sidDefaultZoomSlider"), "change-value",
                    G_CALLBACK(+[](GtkRange*, GtkScrollType, gdouble value, gpointer self) -> gboolean {
                        static_cast<SettingsDialog*>(self)->setDefaultPageZoom(round_cast<int>(value));
                        return false;
                    }),
                    this);
```

**Reply by bhennion** [reply]
And the fact that it works for the zoomCallibSlider callback is random luck. It should be fixed as well.

**Reply by GithubAnon0000** [reply]
Thank you very much! I tested the zoomCallibSlider callback using your pedantic way and it works as expected. Do you want me to open a seperate pr for that?
